server/src/test/java/org/elasticsearch/index/mapper/TextFieldMapperTests.java

```diff
@@ -956,4 +963,101 @@ public void testIndexPrefixMapping() throws IOException {
+            .addClause(new SpanTermQuery(new Term("field", "two")))
+            .addClause(new FieldMaskingSpanQuery(
+                new SpanTermQuery(new Term("field._index_prefix", "words")), "field")
+            )
+            .build();
+        assertThat(q, equalTo(expected));
+
+        Query q2 = new MatchPhrasePrefixQueryBuilder("field", "three words here").toQuery(queryShardContext);
+        expected = new SpanNearQuery.Builder("field", true)
+            .addClause(new SpanTermQuery(new Term("field", "three")))
+            .addClause(new SpanTermQuery(new Term("field", "words")))
+            .addClause(new FieldMaskingSpanQuery(
+                new SpanTermQuery(new Term("field._index_prefix", "here")), "field")
+            )
+            .build();
+        assertThat(q2, equalTo(expected));
+
+        Query q3 = new MatchPhrasePrefixQueryBuilder("field", "two words").slop(1).toQuery(queryShardContext);
+        expected = new SpanNearQuery.Builder("field", true)
+            .setSlop(1)
+            .addClause(new SpanTermQuery(new Term("field", "two")))
+            .addClause(new FieldMaskingSpanQuery(
+                new SpanTermQuery(new Term("field._index_prefix", "words")), "field")
+            )
+            .build();
+        assertThat(q3, equalTo(expected));
+
+        Query q4 = new MatchPhrasePrefixQueryBuilder("field", "singleton").toQuery(queryShardContext);
+        assertThat(q4, is(new SynonymQuery(new Term("field._index_prefix", "singleton"))));
+
+        Query q5 = new MatchPhrasePrefixQueryBuilder("field", "sparkle a stopword").toQuery(queryShardContext);
+        expected = new SpanNearQuery.Builder("field", true)
+            .addClause(new SpanTermQuery(new Term("field", "sparkle")))
+            .addGap(1)
+            .addClause(new FieldMaskingSpanQuery(
+                new SpanTermQuery(new Term("field._index_prefix", "stopword")), "field")
+            )
+            .build();
+        assertThat(q5, equalTo(expected));
+
+        MatchQuery matchQuery = new MatchQuery(queryShardContext);
+        matchQuery.setAnalyzer(new MockSynonymAnalyzer());
+        Query q6 = matchQuery.parse(MatchQuery.Type.PHRASE_PREFIX, "synfield", "motor dogs");
+        expected = new SpanNearQuery.Builder("synfield", true)
+            .addClause(new SpanTermQuery(new Term("synfield", "motor")))
+            .addClause(
+                new SpanOrQuery(
+                    new FieldMaskingSpanQuery(
+                        new SpanTermQuery(new Term("synfield._index_prefix", "dogs")), "synfield"
+                    ),
+                    new FieldMaskingSpanQuery(
+                        new SpanTermQuery(new Term("synfield._index_prefix", "dog")), "synfield"
+                    )
+                )
+            )
+            .build();
+        assertThat(q6, equalTo(expected));
+
+        Query q7 = matchQuery.parse(MatchQuery.Type.PHRASE_PREFIX, "field", "motor d");
```

Comment:
Might be interesting in future to use the wildcard expansion from #36703 here as well?

Reply:
Yes, good idea. I can do that in a follow up.